classes/class-pmpro-members-list-table.php

```diff
@@ -297,28 +310,44 @@ private function sql_table_data( $count = false ) {
 				UNIX_TIMESTAMP(CONVERT_TZ(max(mu.enddate), '+00:00', @@global.time_zone)) as enddate, m.name as membership
 				";
 		}
-			
-		$sqlQuery .= 
+
+		$sqlQuery .=
 			"	
 			FROM $wpdb->users u 
 			LEFT JOIN $wpdb->pmpro_memberships_users mu
 			ON u.ID = mu.user_id
 			LEFT JOIN $wpdb->pmpro_membership_levels m
 			ON mu.membership_id = m.id
 			";
-			
+
 		if ( !empty( $s ) ) {
-			$sqlQuery .= " LEFT JOIN $wpdb->usermeta um ON u.ID = um.user_id ";
+			if ( ! empty( $search_key ) ) {
+				// If there's a colon in the search string, make the search smarter.
+				if( in_array( $search_key, array( 'login', 'nicename', 'email', 'url', 'display_name' ) ) ) {
+					$key_column = 'u.user_' . esc_sql( $search_key );
+					$search_query = " AND $key_column LIKE '%" . esc_sql( $s ) . "%' ";
+				} elseif ( $search_key === 'discount' || $search_key === 'discount_code' || $search_key === 'dc' ) {
+					$user_ids = $wpdb->get_col( "SELECT dcu.user_id FROM $wpdb->pmpro_discount_codes_uses dcu LEFT JOIN $wpdb->pmpro_discount_codes dc ON dcu.code_id = dc.id WHERE dc.code = '" . esc_sql( $s ) . "'" );
+					$search_query = " AND u.ID IN(" . implode( ",", $user_ids ) . ") ";
+				} else {
+					$user_ids = $wpdb->get_col( "SELECT user_id FROM $wpdb->usermeta WHERE meta_key = '" . esc_sql( $search_key ) . "' AND meta_value lIKE '%" . esc_sql( $s ) . "%'" );
+					$search_query = " AND u.ID IN(" . implode( ",", $user_ids ) . ") ";
```

Comment:
I'm not too concerned but even if we think the value will be an array of numbers, we should probably cast to int anyways like `$user_ids = array_map( 'absint', $user_ids );`

As it is right now, there is no escaping happening on the user IDs before it is inserted into the query and that could become problematic later.